ir/src/main/scala/org/scalajs/ir/Names.scala

```diff
@@ -22,37 +22,96 @@ import java.nio.CharBuffer
 import java.nio.charset.CharacterCodingException
 
 object Names {
-  type EncodedName = Array[Byte]
+  /** An immutable name encoded in UTF-8.
+   *
+   *  The contents of an `EncodedName` is guaranteed to be a well-formed UTF-8
+   *  string.
+   *
+   *  @note
+   *    `equals()` and `hashCode()`, along with `==` and `##`, are just as
+   *    broken for `EncodedName` as for `Array`s. Use the methods in the
+   *    companion object instead.
+   */
+  final class EncodedName private (private[ir] val bytes: Array[Byte])
```

Comment:
I like how this gives a neat separation between base encoding and higher level name guarantees. In this line: Have you considered putting this into a separate file (so the utf8 encoding methods would get closer together without increasing noise in this file).

Reply:
Also, consider simply calling this `UTF8String`.

Reply:
Done.